src/metadata.cpp

```diff
@@ -21,95 +21,110 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 #include "log.h"
 
 /*
-	Metadata
+	IMetadata
 */
 
-void Metadata::clear()
+bool IMetadata::operator==(const IMetadata &other) const
 {
-	m_stringvars.clear();
-	m_modified = true;
-}
-
-bool Metadata::empty() const
-{
-	return m_stringvars.empty();
-}
-
-size_t Metadata::size() const
-{
-	return m_stringvars.size();
-}
-
-bool Metadata::contains(const std::string &name) const
-{
-	return m_stringvars.find(name) != m_stringvars.end();
-}
+	StringMap this_map_, other_map_;
+	const StringMap &this_map = getStrings(&this_map_);
+	const StringMap &other_map = other.getStrings(&other_map_);
 
-bool Metadata::operator==(const Metadata &other) const
-{
-	if (size() != other.size())
+	if (this_map.size() != other_map.size())
 		return false;
 
-	for (const auto &sv : m_stringvars) {
-		if (!other.contains(sv.first) || other.getString(sv.first) != sv.second)
+	for (const auto &this_pair : this_map) {
+		const auto &other_pair = other_map.find(this_pair.first);
+		if (other_pair == other_map.cend() || other_pair->second != this_pair.second)
 			return false;
 	}
 
 	return true;
 }
 
-const std::string &Metadata::getString(const std::string &name, u16 recursion) const
+const std::string &IMetadata::getString(const std::string &name, std::string *place,
+		u16 recursion) const
 {
-	StringMap::const_iterator it = m_stringvars.find(name);
-	if (it == m_stringvars.end()) {
+	const std::string *raw = getStringRaw(name, place);
+	if (!raw) {
 		static const std::string empty_string = std::string("");
 		return empty_string;
 	}
 
-	return resolveString(it->second, recursion);
+	return resolveString(*raw, place, recursion);
 }
 
-bool Metadata::getStringToRef(
-		const std::string &name, std::string &str, u16 recursion) const
+bool IMetadata::getStringToRef(const std::string &name,
+		std::string &str, u16 recursion) const
 {
-	StringMap::const_iterator it = m_stringvars.find(name);
-	if (it == m_stringvars.end()) {
+	const std::string *raw = getStringRaw(name, &str);
+	if (!raw)
 		return false;
-	}
 
-	str = resolveString(it->second, recursion);
+	const std::string &resolved = resolveString(*raw, &str, recursion);
```

Comment:
does this work safely if `raw == &str`?

Reply:
Yes, I believe so. Analyzing the implementation of `resolveString`, it seems that the referenced string in the first parameter is never accessed after the string in the second parameter is modified (if it is modified at all).